### PR TITLE
Extract fastlane automation to share it between SDKs

### PR DESCRIPTION
**Pull request by tonidero**
### Description
https://revenuecats.atlassian.net/browse/CSDK-27

In this PR we separate part of the fastlane automation currently in `Fastfile` to a separate `CommonFastfile`. This file will be extracted to its own repo in a future PR.

The idea here is to reuse the automation we currently have on the iOS SDK in all our other SDK repos (as much as possible).

Some of the challenges I had to deal with were: 
- The `bump` lane used to recreate `RevenueCat-Swift.h`, which is an iOS only task. In order to extract `bump`, I had to remove that part from that lane. Instead I added a check to the `release-checks` lane to make sure it was updated.
- The common lanes require some specific configuration per repo. In order to avoid wrapping those lanes in other lanes within the main `Fastfile`, I currently set this data as environment variables. Not a 100% sure about this solution, but I think it should be mostly ok.

### TODO
- [x] Test github interactions (create branch, commit, push, create PR,...). The parameters seem to be correct but I didn't want to create release branches and PRs without alerting anyone. I plan to test these with an android release tomorrow.

### CODE REVIEW

File: fastlane/CommonFastfile [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+  create_new_release_branch(version: new_version_number)
+  replace_version_number(version: new_version_number)
+
+  attach_changelog_to_master(new_version_number)
+
+  commit_updated_files_and_push(version: new_version_number)    
+  
+  create_pull_request(
+    title: "Release/#{new_version_number}",
+    base: "main",
```
Comment (tonidero) [inline fastlane/CommonFastfile:37]: This doesn't really work for hotfixes. But I think this is ok for now, we can add more automation in the future. This is how it worked until now in the iOS repo

File: fastlane/Fastfile [hunk 2]
```diff
@@ -207,104 +164,19 @@ platform :ios do
...
+    begin
+      ensure_git_status_clean
+    rescue => e
+      UI.error("RevenueCat-Swift.h autogenerated file changed. Make sure it's up-to-date before releasing.")
+    end
   end
 
   desc "Creates RevenueCat-Swift.h for a new release"
-  private_lane :compile_autogenerated_header do |options|
+  lane :compile_autogenerated_header do |options|
```
Comment (tonidero) [inline fastlane/Fastfile:179]: I made this public in case a dev wants to recreate `RevenueCat-Swift.h` header, which might be useful if we now catch this as an error during the `release-checks` lane.

File: fastlane/README.md [hunk 3]
```diff
@@ -13,66 +13,90 @@ For _fastlane_ installation instructions, see [Installing _fastlane_](https://do
 
 # Available Actions
 
-### check_pods
+### bump
```
Comment (tonidero) [inline fastlane/README.md:16]: The order of these actions has changed... Not sure if that's a problem. It does include the actions from `CommonFastfile` though.

File: fastlane/CommonFastfile [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+  if new_text.to_s.strip.empty? and not allow_empty
+    UI.user_error!("Missing `new_text` in call to `replace_in`, looking for replacement for #{previous_text} 😵.")
+  end
+  sed_regex = 's|' + previous_text.sub(".", "\\.") + '|' + new_text + '|'
+  backup_extension = '.bck'
+  sh("sed", '-i', backup_extension, sed_regex, path)
+end
+
+def current_version_number
+  File.read("../.version").strip
```
Comment (tonidero) [inline fastlane/CommonFastfile:269]: Note how this checks for the existance of a `.version` file in the root folder. This is not the case in most of our repos. However, I think it's a good pattern to follow to have a common versioning system, since Android, iOS, Flutter,... each have their own versioning systems that would be tricky to abstract.

File: fastlane/Fastfile [hunk 5]
```diff
@@ -10,8 +10,22 @@
...
+ENV["FILES_TO_UPDATE_VERSION_STRING"] = "
+  ../RevenueCat.podspec,
+  ../Sources/Misc/SystemInfo.swift,
+  ../.jazzy.yaml,
+  ../.version,
+  ../Examples/MagicWeather/MagicWeather.xcodeproj/project.pbxproj,
+  ../Examples/MagicWeatherSwiftUI/Magic Weather SwiftUI.xcodeproj/project.pbxproj,
+  ../scripts/docs/index.html
+"
+ENV["GITHUB_RELEASE_UPLOAD_ASSETS"] = "RevenueCat.framework.zip,RevenueCat.xcframework.zip"
```
[the remark this answers is not in the document]
Comment (taquitos) [reply fastlane/Fastfile:27]: Nah, I think that's fine

File: fastlane/CommonFastfile (hunk 4 above)
Comment (taquitos) [reply fastlane/CommonFastfile:269]: 💯💯💯

File: fastlane/CommonFastfile [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+desc "Prepare next version"
+lane :prepare_next_version do |options|
+  old_version_number = current_version_number
+  major, minor, _ = old_version_number.split('.')
+  next_version = "#{major}.#{minor.to_i + 1}.0"
+  next_version_snapshot = "#{next_version}-SNAPSHOT"
+
+  org = "RevenueCat"
+  repo = ENV["REPO_NAME"]
```
[the remark this answers is not in the document]
Comment (taquitos) [reply fastlane/CommonFastfile:50]: can we add a prompt with a selector (and default timeout for no-entry) in the event that the `ENV` doesn't contain anything?

File: fastlane/Fastfile (hunk 2 above)
Comment (taquitos) [reply fastlane/Fastfile:179]: 👍

File: fastlane/Fastfile [hunk 7]
```diff
@@ -644,39 +444,3 @@ end
...
-def replace_in(previous_text, new_text, path, allow_empty=false)
-  if new_text.to_s.strip.empty? and not allow_empty
-    UI.user_error!("Missing `new_text` in call to `replace_in`, looking for replacement for #{previous_text} 😵.")
-  end
-  sed_regex = 's|' + previous_text.sub(".", "\\.") + '|' + new_text + '|'
-  backup_extension = '.bck'
-  sh("sed", '-i', backup_extension, sed_regex, path)
-end
-
-def current_version_number
```
[the remark this answers is not in the document]
Comment (taquitos) [reply fastlane/Fastfile:680]: I don't think they need to be a lane.

File: Tests/InstallationTests/CommonFiles/RevenueCat-Swift.h [hunk 8]
```diff
@@ -215,6 +215,258 @@ typedef unsigned int swift_uint4  __attribute__((__ext_vector_type__(4)));
 # pragma pop_macro("any")
 #endif
 
+
+/// This class is responsible for all explicit attribution APIs as well as subscriber attributes that RevenueCat offers.
+/// The attributes are additional structured information on a user. Since attributes are writable using a public key
+/// they should not be used for managing secure or sensitive information such as subscription status, coins, etc.
+/// Key names starting with “$” are reserved names used by RevenueCat. For a full list of key restrictions refer
+/// <a href="https://docs.revenuecat.com/docs/subscriber-attributes">to our guide</a>
+SWIFT_CLASS_NAMED("Attribution")
```
[the remark this answers is not in the document]
Comment (tonidero) [reply Tests/InstallationTests/CommonFiles/RevenueCat-Swift.h:224]: As part of the changes on this PR, this won't be generated automatically moving forward, but we will have a test on the `release-checks` lane that will fail if this file is outdated (More conversation on that here: https://revenuecat.slack.com/archives/C03HPN9UE80/p1655465010272579). Devs can also generate it with the lane `compile_autogenerated_header`, made available on this PR. That said, going to revert these changes here so they can be added once we make a release.

File: fastlane/CommonFastfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (tonidero) [reply fastlane/CommonFastfile:50]: You mean using `UI.input` to have the dev input the repo name? I can do that, but I wonder if it's necessary... In my mind, these envs would be set in the `Fastfile` on each SDK so devs don't ever need to input anything for these. If you think there is a use case where we could get here without setting these environment variables I can do that though, and in that case we might want to do the same for the other environment variables.

File: fastlane/CommonFastfile [hunk 9]
```diff
@@ -0,0 +1,270 @@
+before_all do
```
[the remark this answers is not in the document]
Comment (tonidero) [reply fastlane/CommonFastfile:1]: Right... You can set it up on your editor of choice, but github doesn't seem to pick it up... I'm planning to move this file to a different common repo in a future PR, and we can call it `Fastfile` then, so I hope at that point we get it.

File: fastlane/Fastfile [hunk 10]
```diff
@@ -66,68 +86,11 @@ platform :ios do
...
   end
 
-  desc "Replace version number in project and supporting files"
-  lane :replace_version_number do |options|
-    new_version_number = options[:version]
-    UI.user_error!("missing version") unless new_version_number
-    previous_version_number = current_version_number
-    
-    version_number_without_prerelease_modifiers = new_version_number.split("-")[0]
-    increment_version_number(version_number: version_number_without_prerelease_modifiers)
```
Comment (tonidero) [inline fastlane/Fastfile:76]: I hadn't noticed this was an iOS only operation. This increments the version in the xcodeproj in a smarter way, using some specialized tools. 

I substituted this with our more "manual" version replacement (changing all occurences of the old version number with the new one). I tested it a few times, and it seems to generate the same changes, but lmk if you don't like that and we can keep this task as sdk-dependant

**Comment by tonidero** [comment]
I tested the `bump` lane on the `purchases-android` repo pointing to these changes. The changes to support this were: https://github.com/RevenueCat/purchases-android/pull/550 and the created release PR looked like: https://github.com/RevenueCat/purchases-android/pull/549

File: Tests/InstallationTests/CommonFiles/RevenueCat-Swift.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Tests/InstallationTests/CommonFiles/RevenueCat-Swift.h]: Hmm why can't we call this lane from `fastlane bump`? 
Considering we always need to update it, we're going to forget so seems like a point of friction when making releases.

File: fastlane/CommonFastfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (taquitos) [reply fastlane/CommonFastfile:50]: I was thinking in the infrequent case that we need to run this on our local machine we don't want to fiddle with ENV stuff. What you describe here, though makes sense, and it seems like we'd never have to manually set these during a manual deploy.

TLDR: my comment can be ignored 😄

File: fastlane/Fastfile (hunk 10 above)
Comment (taquitos) [reply fastlane/Fastfile:76]: A more platform-agnostic approach like you describe makes sense to me. In the future it might even make sure to extract out `ENV["FILES_TO_UPDATE_VERSION_STRING"]` into separate file that gets loaded.

File: Tests/InstallationTests/CommonFiles/RevenueCat-Swift.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (tonidero) [reply Tests/InstallationTests/CommonFiles/RevenueCat-Swift.h]: Right, I wasn't sure about this and asked it over Slack. Basically the idea is to make the `bump` lane platform independent. Other platforms don't need to update that header, so I removed it from that lane. We have some alternatives if we want to have it, like wrapping the `bump` lane in another iOS only lane, or add some kind of placeholder lane to be implemented by each platform... We could also not share the bump lane, but seems like a lost opportunity. Honestly I didn't like those options too much but maybe having a `bump_ios` lane wrapping the `bump` lane would be ok?

[the remark this answers is not in the document]
Comment (NachoSoto) [reply Tests/InstallationTests/CommonFiles/RevenueCat-Swift.h]: > like wrapping the bump lane in another iOS only lane,

Yeah that's exactly what I had in mind. We can share *most* of the code, but some platform specific things will always be needed.

[the remark this answers is not in the document]
Comment (tonidero) [reply Tests/InstallationTests/CommonFiles/RevenueCat-Swift.h]: Sounds good! Will do that tomorrow. Thanks!

File: fastlane/.env.SAMPLE [hunk 11]
```diff
@@ -9,3 +9,6 @@ GITHUB_PULL_REQUEST_API_TOKEN=
 GITHUB_PULL_REQUEST_ASSIGNEES=
 GITHUB_PULL_REQUEST_REVIEWERS=
 GITHUB_PULL_REQUEST_TEAM_REVIEWERS=
+
+# Optional for updating changelog. Can be the same as GITHUB_PULL_REQUEST_API_TOKEN and improves issues with github api rate limits.
+GITHUB_TOKEN=
```
Comment (tonidero) [inline fastlane/.env.SAMPLE:14]: Adding this token avoided the rate limit issues I had when testing this, so I added it here, since it can be useful.

File: fastlane/Fastfile [hunk 12]
```diff
@@ -10,14 +10,40 @@
...
+ENV["GITHUB_RELEASE_UPLOAD_ASSETS"] = "RevenueCat.framework.zip,RevenueCat.xcframework.zip"
+
 platform :ios do
   before_all do
     setup_circle_ci
     update_fastlane
   end
 
+  desc "Bump version, update swift header, edit changelog, and create pull request"
+  lane :bump_ios do |options|
```
Comment (tonidero) [inline fastlane/Fastfile:42]: From the conversation in https://github.com/RevenueCat/purchases-ios/pull/1719#discussion_r904202260.  This new lane uses the `bump` lane which is platform agnostic and adds a new step which will regenerate the `RevenueCat-Swift.h` header and commit and push if there were changes. Note that this will happen on a separate commit from the version bump changes, I added the version number to the commit name though, so it's easier to track changes on that file.

Comment (tonidero) [reply fastlane/Fastfile:42]: That's correct!

**Comment by tonidero** [comment]
@RevenueCat/sdk bump

File: fastlane/CommonFastfile [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+  )
+end
+
+desc "Make github release"
+lane :github_release do |options|
+  release_version = options[:version]
+  UI.user_error!("missing version") unless release_version
+  org = "RevenueCat"
+  repo = ENV["REPO_NAME"]
+  UI.user_error!("missing repo name env") unless repo
```
Comment (NachoSoto) [inline fastlane/CommonFastfile:76]: This is duplicated 3 times.

File: fastlane/CommonFastfile [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+
+  begin
+    changelog = File.read("../CHANGELOG.latest.md")
+  rescue
+    UI.user_error!("Please add a CHANGELOG.latest.md file before calling this lane")
+  end
+  commit_hash = last_git_commit[:commit_hash]
+  puts commit_hash
```
[the remark this answers is not in the document]
Comment (tonidero) [reply fastlane/CommonFastfile:84]: Probably, I will remove it, since I don't think it's that useful 👍

File: Contributing/RELEASING.md [hunk 15]
```diff
@@ -1,6 +1,6 @@
 #### Releasing:
 1. Create a `fastlane/.env` file with your GitHub API token (see `fastlane/.env.SAMPLE`). This will be used to create the PR, so you should use your own token so the PR gets assigned to you. 
-2. Run `bundle exec fastlane bump`
+2. Run `bundle exec fastlane ios bump_ios`
```
[the remark this answers is not in the document]
Comment (tonidero) [reply Contributing/RELEASING.md:3]: I couldn't call the common `bump` lane directly, since they would share the name and end up in an inifinite loop. I thought about a couple options, like calling the terminal fastlane command from the fastfile with `sh` or moving the logic to a different method that can be shared by the common `bump` lane and the ios `bump` lane, but I thought those weren't clean. In the end, I renamed the common `bump` lane to `bump_version_update_changelog_create_pr`. With this, each platform can create their own `bump` lane that is just a wrapper around the `bump_version_update_changelog_create_pr` lane, which I think is fine and if we have to add platform specific logic in the future is probably cleaner.
